htmresearch/algorithms/sparse_net.py

```diff
@@ -0,0 +1,352 @@
+
+
+  def plotBasis(self, filename=None):
+    """
+    Plots the basis functions, reshaped in 2-dimensional arrays.
+
+    This representation makes the most sense for visual input.
+    :param:  filename    (string) Can be provided to save the figure
+    """
+    if np.floor(np.sqrt(self.inputDim)) ** 2 != self.inputDim:
+      print "Basis visualization is not available if inputDim is not a square."
+      return
+
+    dim = int(np.sqrt(self.inputDim))
+
+    if np.floor(np.sqrt(self.outputDim)) ** 2 != self.outputDim:
+      outDimJ = np.sqrt(np.floor(self.outputDim / 2))
+      outDimI = np.floor(self.outputDim / outDimJ)
+      if outDimI > outDimJ:
+        outDimI, outDimJ = outDimJ, outDimI
+    else:
+      outDimI = np.floor(np.sqrt(self.outputDim))
+      outDimJ = outDimI
+
+    outDimI, outDimJ = int(outDimI), int(outDimJ)
+    basis = - np.ones((1 + outDimI * (dim + 1), 1 + outDimJ * (dim + 1)))
+
+    # populate array with basis values
+    k = 0
+    for i in xrange(outDimI):
+      for j in xrange(outDimJ):
+        colorLimit = np.max(np.abs(self.basis[:, k]))
+        mat = np.reshape(self.basis[:, k], (dim, dim)) / colorLimit
+        basis[1 + i * (dim + 1) : 1 + i * (dim + 1) + dim, \
+              1 + j * (dim + 1) : 1 + j * (dim + 1) + dim] = mat
+        k += 1
+
+    plt.figure()
+    plt.subplot(aspect="equal")
+    plt.pcolormesh(basis)
+
+    plt.axis([0, 1 + outDimJ * (dim + 1), 0, 1 + outDimI * (dim + 1)])
+    # remove ticks
+    plt.gca().xaxis.set_major_locator(plt.NullLocator())
+    plt.gca().yaxis.set_major_locator(plt.NullLocator())
+    plt.title("Basis functions for {0}".format(self))
+
+    if filename is not None:
+      plt.savefig(filename)
+
+
+  def _reset(self):
+    """
+    Reinitializes basis functions, iteration number and loss history.
+    """
+    self.basis = np.random.randn(self.inputDim, self.outputDim)
+    self.basis /= np.sqrt(np.sum(self.basis ** 2, axis=0))
+    self._iteration = 0
+    self.losses = dict()
+
+
+  def _learn(self, batch, activations):
+    """
+    Learns a single iteration on the provided batch and activations.
+    :param batch:      (array)  Training batch, of dimension (inputDim,
+                                batchSize)
+    :param activations:(array)  Computed activations, of dimension (outputDim,
+                                batchSize)
+    """
+    batchResiduals = batch - self.basis.dot(activations)
+    loss = np.mean(np.sqrt(np.sum(batchResiduals ** 2, axis=0)))
+    self.losses[self._iteration] = loss
+
+    if self.verbosity >= 2:
+      if self._iteration % self.showEvery == 0:
+        print "At iteration {0}, loss is {1:.3f}".format(self._iteration, loss)
+
+    # update basis
+    gradBasis = batchResiduals.dot(activations.T) / self.batchSize
+    self.basis += self.learningRate * gradBasis
+
+    # normalize basis
+    self.basis /= np.sqrt(np.sum(self.basis ** 2, axis=0))
+
+
+  def _activation(self, input, threshold, thresholdType=None):
```

Comment:
The name of this function is not very descriptive and may be confused with encode (which also gives activation). Maybe to call it thresholdNonlinearity instead?